src/types/wrappers/MaxMinDrivingForceModel.jl

```diff
@@ -0,0 +1,307 @@
+max min -ΔᵣG
+s.t. ΔᵣG = ΔrG⁰ + R T S' ln(C)
+     ΔᵣG ≤ 0
+     ln(Cₗ) ≤ ln(C) ≤ ln(Cᵤ)
+```
+where `ΔrG` are the Gibbs energies dissipated by the reactions, R is the gas
+constant, T is the temperature, S is the stoichiometry of the model, and C is
+the vector of metabolite concentrations (and their respective lower and upper
+bounds).
+
+In case no feasible solution exists, `nothing` is returned.
+
+Since biochemical thermodynamics are assumed, the `proton_ids` and `water_ids`
+*need* to be specified so that they can be ignored in the calculations.
+Effectively this assumes an aqueous environment at constant pH is used.
+
+# Fields
+$(TYPEDFIELDS)
+"""
+Base.@kwdef mutable struct MaxMinDrivingForceModel <: AbstractModelWrapper
+    "A dictionary mapping ΔrG⁰ to reactions."
+    reaction_standard_gibbs_free_energies::Dict{String,Float64} = Dict{String,Float64}()
+
+    "A cycle-free reference flux solution that is used to set the directions of the reactions."
+    flux_solution::Dict{String,Float64} = Dict{String,Float64}()
+
+    "Metabolite ids of protons."
+    proton_ids::Vector{String} = ["h_c", "h_e"]
+
+    "Metabolite ids of water."
+    water_ids::Vector{String} = ["h2o_c", "h2o_e"]
+
+    "A dictionationay mapping metabolite ids to concentrations that are held constant."
+    constant_concentrations::Dict{String,Float64} = Dict{String,Float64}()
+
+    "A dictionary mapping metabolite ids to constant concentration ratios in the form `(m1, m2) = r === m1/m2 = r`."
+    concentration_ratios::Dict{Tuple{String,String},Float64} =
+        Dict{Tuple{String,String},Float64}()
+
+    "Global metabolite concentration lower bound."
+    concentration_lb = 1e-9
+
+    "Global metabolite concentration upper bound."
+    concentration_ub = 100e-3
+
+    "Thermodynamic temperature."
+    T::Float64 = constants.T
+
+    "Real gas constant."
+    R::Float64 = constants.R
+
+    "Tolerance use to distinguish flux carrying reactions from zero flux reactions."
+    small_flux_tol::Float64 = 1e-6
+
+    "Maximum absolute ΔG bound allowed by a reaction."
+    max_dg_bound::Float64 = 1000.0
+
+    "Reaction ids that are ignored internally during thermodynamic calculations. This should include water and proton importers."
+    ignore_reaction_ids::Vector{String} = String[]
+
+    "Inner metabolic model calculations are based on."
+    inner::AbstractMetabolicModel
+end
+
+Accessors.unwrap_model(model::MaxMinDrivingForceModel) = model.inner
+
+"""
+$(TYPEDSIGNATURES)
+
+The variables for max-min driving force analysis are the actual maximum minimum
+driving force of the model, the log metabolite concentrations, and the gibbs
+free energy reaction potentials across each reaction.
+"""
+Accessors.variables(model::MaxMinDrivingForceModel) =
+    ["mmdf"; "log " .* metabolites(model); "ΔG " .* reactions(model)]
+
+"""
+$(TYPEDSIGNATURES)
+
+Helper function that returns the unmangled variable IDs.
+"""
+get_unmangled_variables(model::MaxMinDrivingForceModel) =
```

Comment:
`mmdf_original_variables`

Reply:
(btw might be better shoveled something into `Internal` or so, so that users can't access it)